docs/PowerAuth-SDK-for-Android.md

```diff
@@ -208,6 +265,100 @@ if (result != PowerAuthErrorCodes.PA2Succeed) {
+
+```java
+final String scannedCode = "VVVVV-VVVVV-VVVVV-VTFVA#aGVsbG8.....gd29ybGQ=";
+final Otp otp = OtpUtil.parseFromActivationCode(scannedCode);
+if (otp == null || otp.activationCode == null) {
+    // Invalid code, QR code should contain a signature
+    return;
+}
+```
+
+Note that the signature is only formally validated in the function above. The actual signature verification is done in the activation process, or you can do it on your own:
+
+```java
+final String scannedCode = "VVVVV-VVVVV-VVVVV-VTFVA#aGVsbG8......gd29ybGQ=";
+final Otp otp = OtpUtil.parseFromActivationCode(scannedCode);
+if (otp == null || otp.activationCode == null) { 
+    return; 
+}
+final byte[] codeBytes = otp.activationCode.getBytes(Charset.defaultCharset());
+final byte[] signatureBytes = Base64.decode(otp.activationSignature, Base64.NO_WRAP);
+if (!powerAuthSDK.verifyServerSignedData(codeBytes, signatureBytes, true)) {
+    // Invalid signature
+}
+```
+
+#### Validating entered activation code
+
+To validate an activation code at once, you can call `OtpUtil.validateActivationCode()` function. You have to provide the code, without the signature part. For example:
+
+```java
+boolean isValid   = OtpUtil.validateActivationCode("VVVVV-VVVVV-VVVVV-VTFVA");
+boolean isInvalid = OtpUtil.validateActivationCode("VVVVV-VVVVV-VVVVV-VTFVA#aGVsbG8gd29ybGQ=");
+```
+
+If your application is using your own validation, then you should switch to functions provided by SDK. The reason for that is that since SDK `1.0.0`, all activation codes contains a checksum, so it's possible to detect mistyped characters before you start the activation. Check our [Activation Code](https://github.com/wultra/powerauth-crypto/blob/develop/docs/Activation-Code.md) documentation for more details.
+
+#### Validating recovery code and PUK
+
+To validate a recovery code at once, you can call `OtpUtil.validateRecoveryCode()` function. You can provide the whole code, which may, or may not contain `"R:"` prefix. So, you can validate manually entered codes, but also codes scanned from QR. For example:
+
+```java
+boolean isValid1 = OtpUtil.validateRecoveryCode("VVVVV-VVVVV-VVVVV-VTFVA");
+boolean isValid2 = OtpUtil.validateRecoveryCode("R:VVVVV-VVVVV-VVVVV-VTFVA");
+```
+
+To validate PUK at once, you can call `OtpUtil.validateRecoveryPuk()` function:
+
+```java
+boolean isValid   = OtpUtil.validateRecoveryPuk("0123456789");
+```
+
+#### Auto-correcting typed characters
+
+You can implement auto-correcting of typed characters with using `OtpUtil.validateAndCorrectTypedCharacter()` function in screens, where user suppose to enter an activation, or recovery code. This technique is possible due to fact, that Base32 is specially constructed that doesn't contain visually confusing characters. For example, `1` (number one) and `I` (capital I) are confusing, so only `I` is allowed. The benefit is that provided function can correct typed `1` and translate it to `I`. 
```

Comment:
`where user suppose to enter an activation, or recovery code` -> `where user is supposed to enter an activation or recovery code`